File: src/lib/algorithms/burkes.ts
// Burkes error diffusion dithering
export function burkesDithering(
  grayscale: Uint8ClampedArray,
  width: number,
  height: number,
  threshold: number = 128
): ImageData {
  // Create output ImageData
  const outputData = new ImageData(width, height);
  const output = outputData.data;
  
  // Create a copy of the grayscale image to apply error diffusion
  const buffer = new Float32Array(grayscale);
  
  // Apply dithering
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const idx = y * width + x;
      const oldPixel = buffer[idx];
      
      // Apply threshold
      const newPixel = oldPixel < threshold ? 0 : 255;
      
      // Calculate error
      const error = oldPixel - newPixel;
      
      // Distribute error to neighboring pixels using Burkes pattern
      // Forward row
      if (x + 1 < width) buffer[idx + 1] += error * (8 / 32);
      if (x + 2 < width) buffer[idx + 2] += error * (4 / 32);
      
      // Next row
      if (y + 1 < height) {
        if (x - 2 >= 0) buffer[idx + width - 2] += error * (2 / 32);
        if (x - 1 >= 0) buffer[idx + width - 1] += error * (4 / 32);
        buffer[idx + width] += error * (8 / 32);
        if (x + 1 < width) buffer[idx + width + 1] += error * (4 / 32);
        if (x + 2 < width) buffer[idx + width + 2] += error * (2 / 32);
      }
      
      // Set all RGB channels to the result (black or white)
      output[idx * 4] = newPixel;     // R
      output[idx * 4 + 1] = newPixel; // G
      output[idx * 4 + 2] = newPixel; // B
      output[idx * 4 + 3] = 255;      // A (fully opaque)
    }
  }
  
  return outputData;
} 

File: rust_wasm/src/burkes.rs
pub fn burkes_dither(
    input_ptr: *const u8,
    width: usize,
    height: usize,
    threshold: u8,
    output_ptr: *mut u8,
) {
    let size = width * height;
    let input = unsafe { std::slice::from_raw_parts(input_ptr, size) };
    let output = unsafe { std::slice::from_raw_parts_mut(output_ptr, size * 4) }; // RGBA output

    // Create a mutable buffer for error diffusion
    let mut buffer: Vec<f32> = input.iter().map(|&p| p as f32).collect();

    for y in 0..height {
        for x in 0..width {
            let idx = y * width + x;
            let old_pixel = buffer[idx];

            // Apply threshold
            let new_pixel_value = if old_pixel < threshold as f32 { 0.0 } else { 255.0 };
            let error = old_pixel - new_pixel_value;

            // Write RGBA output pixel
            let base_idx_out = idx * 4;
            output[base_idx_out] = new_pixel_value as u8;
            output[base_idx_out + 1] = new_pixel_value as u8;
            output[base_idx_out + 2] = new_pixel_value as u8;
            output[base_idx_out + 3] = 255; // Alpha

            // Distribute error (Burkes)
            // Current row
            if x + 1 < width {
                buffer[idx + 1] += error * (8.0 / 32.0);
            }
            if x + 2 < width {
                buffer[idx + 2] += error * (4.0 / 32.0);
            }
            // Next row
            if y + 1 < height {
                if x >= 2 {
                    buffer[idx + width - 2] += error * (2.0 / 32.0);
                }
                if x >= 1 {
                    buffer[idx + width - 1] += error * (4.0 / 32.0);
                }
                buffer[idx + width] += error * (8.0 / 32.0);
                if x + 1 < width {
                    buffer[idx + width + 1] += error * (4.0 / 32.0);
                }
                if x + 2 < width {
                    buffer[idx + width + 2] += error * (2.0 / 32.0);
                }
            }
        }
    }
} 